### PR TITLE
Honeywell/Ademco Door/Window Security Sensor driver

### PR DESCRIPTION
**Pull request by adam1010**
I've created a driver for the Honeywell/Ademco Door/Window security sensor.

### CODE REVIEW

**Comment by adam1010** [comment]
Please Squash & Merge as my code was slightly out of date and I needed to make a 2nd commit

File: src/devices/honeywell.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+static int honeywell_callback(bitbuffer_t *bitbuffer) {
+  if(bitbuffer->num_rows == 0)
+    return 0;
+  
+  int good = 0;
+  char time_str[LOCAL_TIME_BUFLEN];
+        local_time_str(0, time_str);
+
+  const uint8_t *bb;
```
Comment (merbanan) [inline src/devices/honeywell.c]: Can you move the declaration of variables to the start of the block ?

File: src/devices/honeywell.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    sprintf(hex, "%02x", bb[3]);
+    device_id[1] = hex[0];
+    device_id[2] = hex[1];
+    sprintf(hex, "%02x", bb[4]);
+    device_id[3] = hex[0];
+    device_id[4] = hex[1];
+    device_id[5] = '\0';
+
+    // Raw Binary
+    char binary [65];
```
Comment (merbanan) [inline src/devices/honeywell.c]: Move declaration to start of block.

File: src/devices/honeywell.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    for (uint16_t bit = 0; bit < 64; ++bit) {
+        if (bb[bit/8] & (0x80 >> (bit % 8))) {
+                binary[bit] = '1';
+        } else {
+                binary[bit] = '0';
+        }
+    }
+    binary[64] = '\0';
+
+    int heartbeat = 0;
```
Comment (merbanan) [inline src/devices/honeywell.c]: Same as before.

File: src/devices/honeywell.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    uint16_t crc_calculated = crc16_buypass(bb);
+    uint16_t crc = (((uint16_t) bb[6]) << 8) + ((uint16_t) bb[7]);
```
Comment (merbanan) [inline src/devices/honeywell.c]: Can't you use the crc functions that already exists ?

File: src/devices/honeywell.c [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    data_t *data = data_make("time",     "", DATA_STRING, time_str,
+                     "time_unix","", DATA_INT, time(NULL),
+                     "id",       "", DATA_STRING, device_id,
+                     "state",    "", DATA_STRING, ( (bb[5] & 0x80) == 0x00)? "closed":"open",
+                     "checksum", "", DATA_STRING, crc == crc_calculated? "ok":"bad",
+                     "heartbeat" , "", DATA_STRING, ( (bb[5] & 0x04) == 0x04)? "yes" : "no",
+                     "binary",   "", DATA_STRING, binary,
+                            NULL);
```
Comment (merbanan) [inline src/devices/honeywell.c]: You most follow the order model that other sensors use, add any extra parameters after those.

**Comment by merbanan** [comment]
Can you send some signal samples for the test repo? After the requested changes the code is ok for merging.

File: src/devices/honeywell.c (hunk 4 above)
Comment (adam1010) [inline src/devices/honeywell.c]: Where would I find those? Does the crc16 one that I included seem to be identical to one in the library?

Comment (merbanan) [inline src/devices/honeywell.c]: https://github.com/merbanan/rtl_433/blob/master/src/util.c

Comment (merbanan) [inline src/devices/honeywell.c]: Look at crc16() and crc16_ccitt()

**Comment by adam1010** [comment]
I've revised the code to put variable declarations at the beginning.

Should the samples be multiple events (door open & door close) from multiple sensors all in one sample?  Or do I need to do Door Open A, Door Close A, Door Open B, Door Close B?

**Comment by merbanan** [comment]
A good set of samples that you can be bothered to collect. I would prefer to not have all types in one file, 2-3 signals in one is ok. Make them related.

File: src/devices/honeywell.c (hunk 4 above)
Comment (adam1010) [inline src/devices/honeywell.c]: crc16_ccitt() was exactly what I needed, thank you!  I've updated the driver file.

File: src/devices/honeywell.c [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                binary[bit] = '1';
+        } else {
+                binary[bit] = '0';
+        }
+    }
+    binary[64] = '\0';
+
+    crc_calculated = crc16_ccitt(bb, 6, 0x8005, 0xfffe);
+    crc = (((uint16_t) bb[6]) << 8) + ((uint16_t) bb[7]);
+
```
Comment (merbanan) [inline src/devices/honeywell.c]: Sorry I missed this, but you are supposed to reject packets with bad crcs.

File: src/devices/honeywell.c [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    data_t *data = data_make(
+                     "time",     "", DATA_STRING, time_str,
+                     "id",       "", DATA_STRING, device_id,
+                     "state",    "", DATA_STRING, ( (bb[5] & 0x80) == 0x00)? "closed":"open",
+                     "heartbeat" , "", DATA_STRING, ( (bb[5] & 0x04) == 0x04)? "yes" : "no",
+                     "checksum", "", DATA_STRING, crc == crc_calculated? "ok":"bad",
```
Comment (merbanan) [inline src/devices/honeywell.c]: This should just be "crc" and a static "ok".

File: src/devices/honeywell.c [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+*
+*/
+
+#include "rtl_433.h"
+#include "pulse_demod.h"
+#include "util.h"
+
+static int honeywell_callback(bitbuffer_t *bitbuffer) {
+  if(bitbuffer->num_rows == 0)
+    return 0;
```
Comment (merbanan) [inline src/devices/honeywell.c]: And this after the variables.

File: src/devices/honeywell.c (hunk 6 above)
Comment (merbanan) [inline src/devices/honeywell.c]: And you should calculate the crc first so you can reject without doing the rest of the code.

File: src/devices/honeywell.c [hunk 9]
```diff
@@ -0,0 +1,98 @@
+/*
+* *** Honeywell (Ademco) Door/Window Sensors ***
+*
+* Tested with the Honeywell 5811 Wireless Door/Window transmitters
+*
```
Comment (merbanan) [inline src/devices/honeywell.c]: If you bother add a tiny description of the bitstream here. You can look at other files for examples.

File: src/devices/honeywell.c [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      }
+  }
+  binary[64] = '\0';
+
+  data_t *data = data_make(
+                   "time",     "", DATA_STRING, time_str,
+                   "id",       "", DATA_STRING, device_id,
+                   "state",    "", DATA_STRING, ( (bb[5] & 0x80) == 0x00)? "closed":"open",
+                   "heartbeat" , "", DATA_STRING, ( (bb[5] & 0x04) == 0x04)? "yes" : "no",
+                   "checksum", "", DATA_STRING, "ok",
```
Comment (adam1010) [inline src/devices/honeywell.c]: All set!

**Comment by merbanan** [comment]
Thanks.

**Comment by panarom** [comment]
thanks very much, @adam1010 : I've been meaning to get these signals decoded for awhile, and I'm delighted to find out somebody else has already take care of it

**Comment by adam1010** [comment]
@panarom Glad it helped! I've been using the code with my home security system for the last year and haven't had any trouble with it. Enjoy!

**Pull request merged**